lib/AST/TS2Flow.cpp

```diff
@@ -0,0 +1,215 @@
+  /// \return true if the program has no TS node, false otherwise.
+  bool validate(ESTree::ProgramNode *program) {
+    auto &sm = context_.getSourceErrorManager();
+    if (sm.getErrorCount())
+      return false;
+
+    visitESTreeNode(*this, program);
+
+    return sm.getErrorCount() == 0;
+  }
+
+  bool incRecursionDepth(ESTree::Node *) {
+    return true;
+  }
+
+  void decRecursionDepth() {}
+
+  void visit(ESTree::Node *node) {
+    auto kind = node->getKind();
+
+    // TSTypeAnnotation is the first TS node and TSModifiers is the last.
+    // This must be in sync with ESTree.def.
+    if (kind >= ESTree::NodeKind::TSTypeAnnotation &&
+        kind <= ESTree::NodeKind::TSModifiers) {
+      return reportError(node);
+    }
+
+    // Update/restore parent_ before/after visiting the children,
+    // so we can keep track of parent_ when we cannot define
+    // `void visit(ESTree::Node *node, ESTree::Node *parent)`.
+    ScopeGuard guard(parent_, node);
+    visitESTreeChildren(*this, node);
+  }
+
+ private:
+  void reportError(ESTree::Node *node) {
+    context_.getSourceErrorManager().error(
+        node->getSourceRange(),
+        "ts2flow: remaining TS node " + node->getNodeName() + " in " +
+            (parent_ ? parent_->getNodeName() : "<unknown>"));
+  }
+
+  Context &context_;
+
+  ESTree::Node *parent_{};
+
+  template <typename T>
+  class ScopeGuard {
+   public:
+    ScopeGuard(T &ref, T value) : ref_(ref), value_(value) {
+      std::swap(ref_, value_);
+    }
+    ~ScopeGuard() {
+      std::swap(ref_, value_);
+    }
+    ScopeGuard(const ScopeGuard &) = delete;
+    ScopeGuard &operator=(const ScopeGuard &) = delete;
+
+   private:
+    T &ref_;
+    T value_;
+  };
+};
+
+class TS2FlowConverter {
+ public:
+  TS2FlowConverter(Context &context)
+      : context_(context), sm_(context.getSourceErrorManager()) {}
+
+  /// Convert any TS annotation nodes to Flow annotation nodes, in place.
+  /// \param program the top-level program node.
+  /// \return the converted program node. nullptr on error.
+  ESTree::ProgramNode *convert(ESTree::ProgramNode *program) {
+    if (sm_.getErrorCount())
+      return nullptr;
+
+    visitESTreeNode(*this, program);
+
+    if (!TS2FlowValidator(context_).validate(program)) {
+      return nullptr;
+    }
+
+    return program;
+  }
+
+  bool incRecursionDepth(ESTree::Node *) {
+    return true;
+  }
+  void decRecursionDepth() {}
+
+  /// Base case: visit the node's children.
+  void visit(ESTree::Node *node) {
+    visitESTreeChildren(*this, node);
+  }
+
+  void visit(ESTree::IdentifierNode *node) {
+    node->_typeAnnotation = convertTypeNode(node->_typeAnnotation);
+    visitESTreeChildren(*this, node);
+  }
+
+  void visit(ESTree::ClassPropertyNode *node) {
+    node->_typeAnnotation = convertTypeNode(node->_typeAnnotation);
+    // Just drop accessibility and readonly modifiers since there is no
+    // counterpart in Flow.
+    node->_tsModifiers = nullptr;
```

Comment:
Private identifiers are actually different from private accessibility in TypeScript, in the sense that private identifiers are runtime feature defined by JS, while TS private accessibility is purely compile time thing. TypeScript also support private identifiers separately: https://www.typescriptlang.org/docs/handbook/2/classes.html#caveats

About reporting error, I thought it might be a bit troublesome if we enforce users to change the source code -- we might want to try some existing TS code as is. How about reporting a warning?

Reply:
True. I just thought we might want to treat them slightly differently since their original runtime behaviours are different.

Would it be okay to ignore the modifiers after reporting a warning (not error) for now?

Reply:
Okay, I will make it report an error.

Just so you know, in my mind, it was as if lexically removing `private` from the source code and thus we don't need to define what its semantics just yet. And that's why I was reluctant to treat private accessibility the same as private identifiers, because doing so is defining the semantics, especially FlowChecker anyway doesn't support it yet. Actually, removing `private` from the source code is what users manually have to do anyway if shermes gives them error. I just thought it might be convenient if the converter does that for them until we know what to do for `private` (or, even `protected`).